tools/diagnose/src/cmd/walk.rs

```diff
@@ -43,9 +55,187 @@ pub enum SymlinkStrategy {
 }
 
 #[tracing::instrument(skip(args), fields(root_device_id))]
-pub fn main(root: PathBuf, args: &WalkArgs) -> Result<()> {
+pub fn main(root: PathBuf, mode: Mode, args: &WalkArgs) -> Result<()> {
+    match mode {
+        Mode::Log => main_logged(root, args),
+        Mode::Tui => {
+            let start = Instant::now();
+            let mut terminal = ratatui::init();
+            let result = main_tui(root, args, &mut terminal);
+            ratatui::restore();
+
+            match result {
+                Ok(stats) => {
+                    println!("walk stats:");
+                    println!("  skipped: {}", stats.skipped);
+                    println!("  warnings: {}", stats.warnings);
+                    println!("  loops: {}", stats.loops);
+                    println!("  visited: {}", stats.visited);
+                    println!("  time: {:?}", start.elapsed());
+                    Ok(())
+                }
+                Err(err) => Err(err),
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+struct WalkStats {
+    skipped: usize,
+    warnings: usize,
+    loops: usize,
+    visited: usize,
+}
+
+#[tracing::instrument(skip(args), fields(root_device_id))]
+fn main_tui(root: PathBuf, args: &WalkArgs, terminal: &mut DefaultTerminal) -> Result<WalkStats> {
+    let regexes = args.filters.iter().map(|rx| rx.as_str());
+    let filters = RegexSet::new(regexes).wrap_err("parse provided regular expressions as a set")?;
+
+    let mut skipped_count = 0usize;
+    let mut warnings_count = 0usize;
+    let mut loops_count = 0usize;
+    let mut visited_count = 0usize;
+    let mut skipped = BoundedVecDeque::new(1000);
+    let mut warnings = BoundedVecDeque::new(1000);
+    let mut loops = BoundedVecDeque::new(1000);
+    let mut visited = BoundedVecDeque::new(1000);
+    let mut drawn = Instant::now();
+
+    const FPS: u64 = 30;
+    const TIME_PER_FRAME: Duration = Duration::from_millis(1000 / FPS);
+
+    let walker = WalkDir::new(&root)
+        .follow_links(matches!(args.symlinks, SymlinkStrategy::Follow))
+        .into_iter();
+
+    for entry in walker {
+        let path = match entry {
+            Ok(entry) => entry.path().to_string_lossy().to_string(),
+            Err(err) => {
+                let msg = format!("{err:#}");
+                if msg.contains("loop") {
+                    loops.push_front(msg);
+                    loops_count += 1;
+                } else {
+                    warnings.push_front(msg);
+                    warnings_count += 1;
+                }
+
+                continue;
+            }
+        };
+
+        if filters.is_match(&path) {
+            skipped.push_front(path.to_string());
```

Comment:
[nit] Isn't `path` already a string here?

Reply:
no, it's a `Cow<'_, str>`